features/netsocket/SocketStats.h

```diff
@@ -0,0 +1,115 @@
+ * limitations under the License.
+ */
+
+#ifndef SOCKET_STATS_H
+#define SOCKET_STATS_H
+
+#include "platform/mbed_stats.h"
+#include "platform/SingletonPtr.h"
+#include "platform/PlatformMutex.h"
+#include "netsocket/Socket.h"
+#include "SocketAddress.h"
+#include "hal/ticker_api.h"
+
+#ifndef MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT
+#define MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT      10
+#endif
+
+typedef enum {
+    SOCK_CLOSED,                    /**< Socket is closed and does not exist anymore in the system */
+    SOCK_OPEN,                      /**< Socket is open, but not associated to any peer address */
+    SOCK_CONNECTED,                 /**< Socket is associated to peer address, either by connect() or sendto()/recvfrom() calls */
+    SOCK_LISTEN,                    /**< Socket is listening for incoming connections */
+} socket_state;
+
+typedef struct {
+    Socket *reference_id;           /**< Used for identifying socket */
+    SocketAddress peer;             /**< Last associated peername of this socket (Destination address) */
+    socket_state state;             /**< State of this socket */
+    nsapi_protocol_t proto;         /**< Specifies a protocol used with socket */
+    size_t sent_bytes;              /**< Data sent through this socket */
+    size_t recv_bytes;              /**< Data received through this socket */
+    us_timestamp_t last_change_tick;/**< osKernelGetTick() when state last changed */
+} mbed_stats_socket_t;
+
+/**  SocketStats class
+ *
+ *   Class to get the network socket statistics
+ */
+class SocketStats {
+public:
+
+    /** Create an socket statictics object
+     *
+     *  Applicatiopn users must not create class object.
+     *  Class object will be created by entities reporting network statistics.
+     *  Application can fetch network statistics using static `mbed_stats_socket_get_each` API
+     *  without creatin an object.
+     */
+    SocketStats();
+    virtual ~SocketStats()
+    {
+    }
+
+    /**
+     *  Fill the passed array of structures with the socket statistics for each created socket.
+     *
+     *  @param stats    A pointer to an array of mbed_stats_socket_t structures to fill
+     *  @param count    The number of mbed_stats_socket_t structures in the provided array
+     *  @return         The number of mbed_stats_socket_t structures that have been filled.
+     *                  If the number of sockets on the system is less than or equal to count,
+     *                  it will equal the number of sockets created (active / closed).
+     *                  If the number of sockets on the system is greater than count,
+     *                  it will equal count.
+     */
+    static size_t mbed_stats_socket_get_each(mbed_stats_socket_t *stats, size_t count);
+
+    /** Add entry of newly created socket in statistics array.
+        @Note: Entry in the array will be maintained even after socket is closed.
```

Comment:
Multi-line editing is being a pain ->

>Entry in the array will be maintained even after socket is closed. It will be over-written for closed sockets when socket entries in `MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT` exceed.

to

>Entries in the array will be maintained even after the socket is closed. Closed sockets will be overwritten first in the case of `MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT` being exceeded.